platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterUtils.java

```diff
@@ -49,7 +70,118 @@ public static String formatNodeName(final long nodeId) {
+            rosterEntries.add(rosterEntry);
+        }
+        return Roster.newBuilder().rosterEntries(rosterEntries).build();
+    }
+
+    /**
+     * Converts an address to a roster entry.
+     *
+     * @param address the address to convert
+     * @param nodeId  the node ID to use for the roster entry
+     * @return the equivalent roster entry
+     */
+    private static RosterEntry toRosterEntry(@NonNull final Address address, @NonNull final NodeId nodeId) {
+        Objects.requireNonNull(address);
+        Objects.requireNonNull(nodeId);
+        final var signingCertificate = address.getSigCert();
+        Bytes signingCertificateBytes;
+        try {
+            signingCertificateBytes =
+                    signingCertificate == null ? Bytes.EMPTY : Bytes.wrap(signingCertificate.getEncoded());
+        } catch (final CertificateEncodingException e) {
+            signingCertificateBytes = Bytes.EMPTY;
+        }
+
+        final List<ServiceEndpoint> serviceEndpoints = new ArrayList<>(2);
+        if (address.getHostnameInternal() != null) {
+            serviceEndpoints.add(endpointFor(address.getHostnameInternal(), address.getPortInternal()));
+        }
+        if (address.getHostnameExternal() != null) {
+            serviceEndpoints.add(endpointFor(address.getHostnameExternal(), address.getPortExternal()));
+        }
+
+        return RosterEntry.newBuilder()
+                .nodeId(nodeId.id())
+                .weight(address.getWeight())
+                .gossipCaCertificate(signingCertificateBytes)
+                .gossipEndpoint(serviceEndpoints)
+                .build();
+    }
+
+    /**
+     * Returns a roster based on the given software upgrade flag,
+     *  the candidate roster, or the active roster from the state.
+     *
+     * @param rosterStore   the roster store to use
+     * @param softwareUpgrade the software upgrade flag
+     * @param candidateRoster the candidate roster stored in the state
+     * @param addressBook     the address book used by the network
+     * @param latestConsensusRound the latest consensus round
+     * @return the roster
+     */
+    @NonNull
+    private static Roster determineRoster(
+            final WritableRosterStore rosterStore,
+            final boolean softwareUpgrade,
+            final Roster candidateRoster,
+            final AddressBook addressBook,
+            final long latestConsensusRound) {
+        if (!softwareUpgrade || candidateRoster == null) {
+            final Roster previousActiveRoster = rosterStore.getActiveRoster();
+            // not in software upgrade mode, return the previous active roster if available
+            // otherwise, create a new roster from the address book
+            return Objects.requireNonNullElseGet(previousActiveRoster, () -> createRoster(addressBook));
+        }
+
+        // software upgrade is detected, we adopt the candidate roster
+        rosterStore.adoptCandidateRoster(latestConsensusRound + 1);
```

Comment:
Why is the roster being adopted here? This method, if you trace it back, is being called by the application to get a roster to pass to the platform builder. I thought the platform would take that roster set in the platform builder and set it as the active roster to be used starting in the next round.

Reply:
Will adoption of the `config.txt` still work with the current workflow until we are ready to switch to this new candidate roster adoption logic?

Reply:
@poulok I think our goal is that the Platform doesn't do any state commits under the new design.    This would perform the state change before giving the roster to the platform, which is what we want.   

But the state change is buried inside the utility class.  I think it needs to be explicit at the top level startup sequence.   When TSS takes over with setting the roster, the location of rotating may need to change and that might be harder to do with it buried inside this utility method.

Reply:
moved to the Modifier and Accessor interfaces.